### ISSUE TITLE
error: MultiGraphQualifierHierarchy found an unqualified type

### ISSUE THREAD

**Issue by wmdietlGC** [typetools/checker-framework#1415]
In a repo with master and #1387 take:

```java
class Optional<T> {
  static <T> Optional<T> absent() { return null; }
  static <T> Optional<T> of(T p) { return null; }
}

class Box<T> {
  void box(T p) {}
}

class Crash9 {
  <T extends Enum<T>> void foo(boolean b,
      Box<Optional<T>> box,
      Class<T> enumClass) {
    box.box(
        b
        ? Optional.<T>absent()
        : Optional.of(Enum.valueOf(enumClass, "hi"))
            );
  }
}
```

Running the Nullness Checker results in a few exceptions and assertion errors, starting with:

```
error: MultiGraphQualifierHierarchy found an unqualified type.  Please ensure that your implicit rules cover all cases and/or use a @DefaultQualifierInHierarchy annotation.
  Compilation unit: Crash9.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:727)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:25)
  org.checkerframework.framework.util.MultiGraphQualifierHierarchy.checkAnnoInGraph(MultiGraphQualifierHierarchy.java:506)
  org.checkerframework.framework.util.MultiGraphQualifierHierarchy.isSubtype(MultiGraphQualifierHierarchy.java:470)
  org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory$KeyForQualifierHierarchy.isSubtype(KeyForAnnotatedTypeFactory.java:253)
  org.checkerframework.framework.util.AtmLubVisitor.lubPrimaryOnBoundedType(AtmLubVisitor.java:329)
  org.checkerframework.framework.util.AtmLubVisitor.visitTypevar_Typevar(AtmLubVisitor.java:290)
  org.checkerframework.framework.util.AtmLubVisitor.visitTypevar_Typevar(AtmLubVisitor.java:31)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:662)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:59)
  org.checkerframework.framework.util.AtmLubVisitor.lubTypeArgument(AtmLubVisitor.java:243)
  org.checkerframework.framework.util.AtmLubVisitor.visitDeclared_Declared(AtmLubVisitor.java:184)
  org.checkerframework.framework.util.AtmLubVisitor.visitDeclared_Declared(AtmLubVisitor.java:31)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:59)
  org.checkerframework.framework.util.AtmLubVisitor.visitTypevar_Typevar(AtmLubVisitor.java:287)
  org.checkerframework.framework.util.AtmLubVisitor.visitTypevar_Typevar(AtmLubVisitor.java:31)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:662)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:59)
  org.checkerframework.framework.util.AtmLubVisitor.lubTypeArgument(AtmLubVisitor.java:243)
  org.checkerframework.framework.util.AtmLubVisitor.visitDeclared_Declared(AtmLubVisitor.java:184)
  org.checkerframework.framework.util.AtmLubVisitor.visitDeclared_Declared(AtmLubVisitor.java:31)
  org.checkerframework.framework.util.AtmCombo.accept(AtmCombo.java:308)
  org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor.visit(AbstractAtmComboVisitor.java:59)
  org.checkerframework.framework.util.AtmLubVisitor.lub(AtmLubVisitor.java:66)
  org.checkerframework.framework.util.AnnotatedTypes.leastUpperBound(AnnotatedTypes.java:629)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitConditionalExpression(TypeFromExpressionVisitor.java:105)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitConditionalExpression(TypeFromExpressionVisitor.java:45)
  com.sun.tools.javac.tree.JCTree$JCConditional.accept(JCTree.java:1245)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:36)
```

(This is Crash9... I'm still trying to repro Crash8 and started working on this one...)

**Comment by smillst** [comment]
Fixed in #1387.  However, the test case issues false positive warnings:

```
tests/all-systems/java8inference/Issue1415.java:22: error: [argument.type.incompatible] incompatible types in argument.
            box.box(b ? Optional.<T>absent() : Optional.of(Enum.valueOf(enumClass, "hi")));
                      ^
  found   : @Initialized @NonNull Optional<T extends @Initialized @Nullable Enum<T extends @Initialized @NonNull Enum<T>>>
  required: @Initialized @NonNull Optional<T extends @Initialized @NonNull Enum<T extends @Initialized @NonNull Enum<T>>>
tests/all-systems/java8inference/Issue1415.java:22: error: [conditional.type.incompatible] incompatible types in conditional expression.
            box.box(b ? Optional.<T>absent() : Optional.of(Enum.valueOf(enumClass, "hi")));
                                          ^
  found   : @Initialized @NonNull Optional<T extends @Initialized @NonNull Enum<T extends @Initialized @NonNull Enum<T>>>
  required: @Initialized @NonNull Optional<T extends @Initialized @Nullable Enum<T extends @Initialized @NonNull Enum<T>>>
```

**Comment by wmdietlGC** [comment]
Thanks for the quick fix.
For `java.util.Optional` in particular, we'll want:

```java
import org.checkerframework.checker.nullness.qual.NonNull;

class Optional<T> {
  static <T> Optional<@NonNull T> absent() { return null; }
  static <T> Optional<@NonNull T> of(T p) { return null; }
}
```

so you could use those signatures to get rid of those warnings.

This fix worked for me. I'm still trying to reproduce 2 more problems and will file more issues.
